File: server/TWS Foundation/TWS Business/Entities/DriverCommon.cs
using System.ComponentModel.DataAnnotations;

using CSM_Foundation.Database.Bases;
using CSM_Foundation.Database.Validators;

using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace TWS_Business.Entities;

/// <summary>
///     [Entity] that represent common information for [Drivers] (<see cref="Driver"/> / <see cref="DriverExternal"/>).
/// </summary>
public class DriverCommon
    : TWSEntity {

    #region Properties

    /// <summary>
    ///     Licence identification number.
    /// </summary>
    [StringLength(12, MinimumLength = 8)]
    public string License { get; set; } = null!;

    #endregion

    #region Relations

    /// <summary>
    ///     <see cref="Entities.Situation"/> information.
    /// </summary>
    public Situation Situation { get; set; } = default!;

    /// <summary>
    ///     <see cref="Entities.Status"/> information.
    /// </summary>
    public Status Status { get; set; } = default!;

    /// <summary>
    ///     <see cref="Driver"/> information.
    /// </summary>
    public Driver? Internal { get; set; }

    /// <summary>
    ///     <see cref="DriverExternal"/> information.
    /// </summary>
    public DriverExternal? External { get; set; }

    #endregion

    #region Custom Getters 

    /// <summary>
    ///     Gets the [Driver] displayable name.
    /// </summary>
    /// <remarks>
    ///     Needs loaded <see cref="Internal"/> then <see cref="Driver.Employee"/> then <see cref="Employees.Employee.Identification"/>.
    ///     Needs loaded <see cref="External"/> then <see cref="DriverExternal.Identification"/>.
    /// </remarks>
    public string? Name {
        get {
            Identification? ident;

            if (Internal != null) {
                ident = Internal.Employee?.Identification;
            } else {
                ident = External?.Identification;
            }

            if (ident == null)
                return null;

            return $"{ident.Name} {ident.Lastname}";
        }
    }

    #endregion


    protected override (string Property, IValidator[])[] Validations((string Property, IValidator[])[] Container) {
        return [
            ..Container,
            (nameof(License), [ new LengthValidator(8, 12)]),
        ];
    }

    protected override void DesignEntity(EntityTypeBuilder etBuilder) {
        etBuilder.ToTable("Drivers_Commons");

        etBuilder.Property(nameof(License)).HasMaxLength(12);

        etBuilder.Link<DriverCommon, Situation>(
                nameof(Situation),
                nameof(Situation.Drivers),
                Required: true,
                Auto: true
            );
        etBuilder.Link<DriverCommon, Status>(
                nameof(Status),
                nameof(Situation.Drivers),
                Required: true,
                Auto: true
            );
    }
}

File: client/tws_foundation_client/lib/src/services/business/drivers/driver_common.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

/// {entity} class.
///
/// Represents a common entity model for [Driver] and [DriverExternal].
final class DriverCommon extends EntityB<DriverCommon> {
  /// [DriverCommon.license] property key access for [DataMap].
  static const String kLicense = "license";

  //! --> Properties

  /// License identification number.
  ///
  /// Rules >
  ///   1. 13 > length > 9
  String license = "";

  //! <-- Properties

  //! --> Relations

  /// [Status] information.
  Status status = Status();

  /// [Situation] information.
  Situation situation = Situation();

  /// [Driver] information.
  Driver? internal;

  /// [DriverExternal] information.
  DriverExternal? external;

  //! <-- Relations

  //! --> Getters

  /// Gets the {driver} name.
  String? get name {
    Identification? ident;

    if (internal != null) {
      ident = internal?.employee.identification;
    } else {
      ident = external?.identification;
    }

    if (ident == null) return null;

    return '${ident.name} ${ident.lastName}';
  }

  //! <-- Getters

  /// Creates a new [DriverCommon] instance.
  DriverCommon();

  @override
  void decode(DataMap encode) {
    license = encode.get(kLicense);

    DataMap dmStatus = encode.get(FoundationCommonPropertyKeys.kStatus);
    status.decode(dmStatus);

    DataMap dmSituation = encode.get(FoundationCommonPropertyKeys.kSituation);
    situation.decode(dmSituation);

    super.decode(encode);
  }

  @override
  DataMap encode([DataMap? entityObject]) {
    return super.encode(
      <String, Object?>{
        kLicense: license,
        FoundationCommonPropertyKeys.kStatus: status.encode(),
        FoundationCommonPropertyKeys.kSituation: situation.encode(),
      },
    );
  }

  @override
  List<EntityInvalidation<DriverCommon>> evaluate() {
    final List<EntityInvalidation<DriverCommon>> invalidations = <EntityInvalidation<DriverCommon>>[];

    if (license.length > 12 || license.length < 8) {
      invalidations.add(
        EntityInvalidation<DriverCommon>(
          this,
          PropertyInfo(kLicense, String, license),
          'Wrong length ${license.length}',
          '13 > length > 9',
        ),
      );
    }

    return invalidations;
  }
}
